hazelcast/src/main/java/com/hazelcast/internal/partition/operation/PromotionCommitOperation.java

```diff
@@ -265,11 +267,23 @@ private void complete() {
     }
 
     /** Reruns this operation with next {@link #runStage}*/
-    private void scheduleNextRun(RunStage nextState) {
-        runStage = nextState;
+    private void scheduleNextRun(RunStage nextStage) {
+        runStage = nextStage;
+        // since this operation is being rescheduled, its invocation time and call timeout
+        // need to be recalculated (instead of allocating & scheduling a clone)
+        resetCallTimeout();
         getNodeEngine().getOperationService().execute(this);
     }
 
+    private void resetCallTimeout() {
```

Comment:
Could be an option to let the promotion fail? If we are giving this operation a potentially infinite timeout, can this be seen to block migration thread indefinitely?

Reply:
I don't think there's a risk of this operation running indefinitely, we are only resetting the timeout each time we advance to the next stage of execution, so there is progress.

Reply:
makes sense.